### PR TITLE
First after repartition returns the wrong value

### PR DESCRIPTION
**Pull request by boazmohar**
- Added repartition to bolt spark array
- Added a check in first() that the RDD is ordered

### CODE REVIEW

**Comment by freeman-lab** [comment]
@boazmohar nice! Want to add some minimal tests for this behavior as well? Would probably make more sense having the tests here than in `thunder`.

We can also continue the conversation of whether to make `sorted` a keyword arg or not. My main reasons for forcing the sort are (1) it better matches the semantics of the array (2) given the operations that lead to unsorting it won't slow things down much in practice and (3) you can always call `array.tordd().first()` to get the unsorted behavior.

**Comment by jwittenbach** [comment]
My thought was that we have a bit of a pattern with methods on the `BoltArray`

- if a function is a NumPy function, then it performs the same array operation
- if a function is a Spark function, the it performs the same RDD operation

I guess I'm OK with breaking this pattern in the direction of making everything more array-oriented

**Comment by boazmohar** [comment]
@jwittenbach @freeman-lab  So why not sort on `repartition` so that `first` would be responsive as you expected it to be?

File: bolt/spark/array.py [hunk 1]
```diff
@@ -46,6 +46,19 @@ def unpersist(self):
...
+        """
+        Repartitions the underlying RDD
+
+        Parameters
+        ----------
+        npartitions : int
+            Number of partitions to repartion the underlying RDD to
+        """
+
+        self._rdd.repartition(npartitions=npartitions)
```
Comment (jwittenbach) [inline bolt/spark/array.py]: I don't think `repartition` doesn't work through a side-effect. You need to catch the repartitions array from the return value.

```python
self._rdd = self._rdd.repartition(npartitions=npartitions)
return self
```

**Comment by jwittenbach** [comment]
@boazmohar `sortByKey` actually kicks off a job, so we want to wait until either a `first` or a `collect` so that everything can be lazy up until that point. Also, if people do multiple operations that break the ordering, we don't want to have multiple unnecessary `sortByKey` operations happening.

**Comment by freeman-lab** [comment]
@boazmohar the reason not to sort on `repartition` right away is that if you follow with a `swap` or `transpose` or another operation that induces a shuffle, it's much cheaper to only do the sort once at the end instead of doing it twice.

**Comment by freeman-lab** [comment]
@jwittenbach hehe we just said the same thing 😄

**Comment by boazmohar** [comment]
@jwittenbach @freeman-lab Yep, I thought `repartition` also kicks off a job.
Makes sense now.
I will correct the code and add tests.

File: bolt/spark/array.py [hunk 2]
```diff
@@ -46,6 +46,19 @@ def unpersist(self):
...
+        Repartitions the underlying RDD
+
+        Parameters
+        ----------
+        npartitions : int
+            Number of partitions to repartion the underlying RDD to
+        """
+
+        self._rdd = self._rdd.repartition(npartitions)
+        self._ordered = False
```
Comment (jwittenbach) [inline bolt/spark/array.py]: make sure to `return self` here too

Comment (freeman-lab) [inline bolt/spark/array.py]: yeah I sort of prefer @jwittenbach 's version

File: test/spark/test_spark_basic.py [hunk 3]
```diff
@@ -57,6 +57,14 @@ def test_cache(sc):
     b.unpersist()
     assert not b._rdd.is_cached
 
+def test_repartition(sc):
```
Comment (freeman-lab) [inline test/spark/test_spark_basic.py]: great test

**Comment by freeman-lab** [comment]
looks good to me! 👍

**Pull request merged**